### DISCUSSION TITLE
Boundary conditions ignored by MOOSE

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#18149] [Q&A General] [answered]
I created a simple mesh with gmsh for a simple flow under a dam. This is the content of the input file:

https://pastebin.com/tGFruT1s

Yet, when I run the simulation, the output seems to completely ignore the BCs, and I have flow_pot = 0 everywhere including at the boundaries I specified. I made my own custom kernels for this to try to learn how MOOSE works, by the way. And they compile fine:

Material Class: https://pastebin.com/pnUfeM3B
Kernel: https://pastebin.com/HBcT2KBq

.inp mesh file created:
https://pastebin.com/QrY55VzP

**Comment by GiudGiud** [comment, marked as answer]
Hello

Gave a quick look and:
- you dont need to specify zero Neumann BCs on the wall, that is a default
- Is this just a diffusion problem? There are pre-built kernels for this btw
- You dont seem to be initializing permeability in that source file. You only declare it. If it's a constant, you could use a GenericConstantMaterial. You could also have a look at that file to see how the prop_values are retrieved from the input file and placed in the properties at the quadrature points

Guillaume

**Reply by richmondodufisan** [reply]
Hi,

Yes it is, but I was trying to learn how to create the kernel using the material interface so I can apply it to a more complex problem.

I believe I fixed that now, and the simulation does seem to be calculating residuals now unlike before (where it would converge with one iteration and |r|=0). However, when I view the output with paraview, it's still giving me a value of flow_pot = 0 everywhere. I still used the 'ADMaterialProperty' btw, because the problem I eventually want to do isn't just a generic constant. 
Here's the updated Material source file:

https://pastebin.com/stmsL7fK

So I set the _permeability at the quadrature points to always equal the constant value retrieved from the user (with _the_permeability).  The input file and kernel remain unchanged.

**Reply by GiudGiud** [reply]
That's weird. Is paraview showing the initial condition by chance?

If there s a non zero residual, chances are the variable is not identically 0.

**Reply by richmondodufisan** [reply]
No, not even showing the boundary condition. The value of flow_pot ranges from 0 to 1e-38 (which I assume is computer precision for 0)

https://ibb.co/ZJDhQYW

However the residual calculation starts off large so its clearly calculating something:

 0 Nonlinear |R| = 1.300704e+02
      0 Linear |R| = 1.300704e+02
      1 Linear |R| = 1.451357e-05
 1 Nonlinear |R| = 1.031623e-05
      0 Linear |R| = 1.031623e-05
      1 Linear |R| = 1.702694e-12
 2 Nonlinear |R| = 8.267341e-14
 Solve Converged!

**Reply by GiudGiud** [reply]
Ok I cant access the page, my chrome wont let me. Can you please post a paraview screenshot?
Alternatively, could you add postprocessors to characterize your solution? So like ElementAverageValue and SideAverageValue on the boundaries.

I would try to isolate the problem by replacing each part of your custom simulation with generic moose objects and see what improves it.

**Reply by richmondodufisan** [reply]
![parav](https://user-images.githubusercontent.com/52802063/123199022-f52a3680-d462-11eb-8236-1bc751b16a9e.JPG)

This is the paraview screenshot.

I added the postprocessors and its clearly calculating something, the SideAverageValue on each boundary returned the value of the boundary conditions. I also tried "SideAverageValue" for the walls, and that gave me a number in between both boundaries. So it means something is wrong with the .e output file.

**Reply by GiudGiud** [reply]
Did you change anything about the output block?
In paraview, did you make sure to select the last timestep?

**Reply by richmondodufisan** [reply]
This is the output block:

[Outputs]
  exodus = true
  perf_graph = true
[]

And in paraview, for the last timestep (only two timesteps as expected for a Steady State calculation) the whole mesh = 1.2e-38


Edit: I've also been trying PointValue for random points within the mesh and they all return values between the BCs

**Reply by richmondodufisan** [reply]
![parav1](https://user-images.githubusercontent.com/52802063/123202343-43dacf00-d469-11eb-9d42-e6ded5ef6863.JPG)

**Reply by richmondodufisan** [reply]
Edit: I installed the latest version of paraview and finally got the solution displayed. A little bizarre imo but at least its working now. Thank you for your help!

![Capture](https://user-images.githubusercontent.com/52802063/123204418-f06a8000-d46c-11eb-8a4c-46a8f4009fcf.JPG)

**Reply by GiudGiud** [reply]
ah great I was getting super confused. Thanks for letting us know. Good luck for the rest of the model!

**Comment by Traiwit** [comment]
Hi @GiudGiud 

I see you mentioned 

> 'You dont seem to be initializing permeability in that source file. You only declare it. If it's a constant, you could use a GenericConstantMaterial. You could also have a look at that file to see how the prop_values are retrieved from the input file and placed in the properties at the quadrature points'

I wonder if we can pre-define material properties such as permeability/porosity in the mesh file (.inp) and MOOSE will recognise it? 

I want to specify a different value of the property at each element (base on element number/ID), is this possible with MOOSE?

Thank you.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
I'm not aware of this capability for .inp files.
Input of material properties is usually done per block.
You should post in another discussions thread to find out if there's a per-element option. I dont think there is, but maybe someone has done something similar before.
